dali/operators/util/get_property.cc

```diff
@@ -38,6 +39,110 @@ The following properties are supported:
+    auto shape = sample_shape(in, i);
+    if (i == 0)
+      tls.resize(N, shape.sample_dim());
+    tls.set_tensor_shape(i, shape);
+  }
+  out.Resize(tls, DALI_UINT8);
+  for (int i = 0; i < N; i++) {
+    copy_sample(out, in, i);
+  }
+}
+
+template <typename Backend>
+void SourceInfoToTL(TensorList<CPUBackend> &out, const TensorList<Backend> &in) {
+  GetPerSample(out, in,
+    [](auto &in, int idx) {
+      auto &info = in.GetMeta(idx).GetSourceInfo();
+      return TensorShape<1>(info.length());
+    },
+    [](auto &out, auto &in, int idx) {
+      auto &info = in.GetMeta(idx).GetSourceInfo();
+      std::memcpy(out.raw_mutable_tensor(idx), info.c_str(), info.length());
+    });
+}
+
+template <typename Backend>
+void SourceInfoToTL(TensorList<GPUBackend> &out, const TensorList<Backend> &in) {
+  TensorList<CPUBackend> tmp;
+  tmp.set_pinned(true);
+  SourceInfoToTL(tmp, in);
+  tmp.set_order(out.order());
+  out.Copy(tmp);
+}
+
+template <typename OutputBackend>
+void SourceInfoToTL(TensorList<OutputBackend> &out, const Workspace &ws) {
+  ws.Output<OutputBackend>(0).set_order(ws.output_order());
+  if (ws.InputIsType<CPUBackend>(0))
+    return SourceInfoToTL(out, ws.Input<CPUBackend>(0));
+  else if (ws.InputIsType<GPUBackend>(0))
+    return SourceInfoToTL(out, ws.Input<GPUBackend>(0));
+  else
+    DALI_FAIL("Internal error - input 0 is neither CPU nor GPU.");
+}
+
+template <typename Backend>
+void RepeatTensor(TensorList<Backend> &tl, const Tensor<Backend> &t, int N) {
+  tl.Reset();
+  tl.set_device_id(t.device_id());
+  tl.SetSize(N);
+  tl.set_sample_dim(t.ndim());
+  tl.set_type(t.type());
+  tl.SetLayout(t.GetLayout());
+  for (int i = 0; i < N; i++)
+    tl.SetSample(i, t);
+}
+
+template <typename Backend>
+void RepeatFirstSample(TensorList<Backend> &tl, int N) {
+  Tensor<Backend> t;
+  TensorShape<> shape = tl[0].shape();
+  t.ShareData(unsafe_sample_owner(tl, 0), shape.num_elements(), tl.is_pinned(),
+              shape, tl.type(), tl.device_id(), tl.order());
+  t.SetMeta(tl.GetMeta(0));
+  RepeatTensor(tl, t, N);
+}
+
+void LayoutToTL(TensorList<CPUBackend> &out, const Workspace &ws) {
+  TensorLayout l = ws.GetInputLayout(0);
+  out.Resize(uniform_list_shape(1, { l.size() }), DALI_UINT8);
+  memcpy(out.raw_mutable_tensor(0), l.data(), l.size());
+  RepeatFirstSample(out, ws.GetInputBatchSize(0));
+}
+
+void LayoutToTL(TensorList<GPUBackend> &out, const Workspace &ws) {
+  TensorLayout l = ws.GetInputLayout(0);
+  Tensor<CPUBackend> tmp_cpu;
+  Tensor<GPUBackend> tmp_gpu;
+  tmp_cpu.Resize(TensorShape<1>(l.size()), DALI_UINT8);
+  memcpy(tmp_cpu.raw_mutable_data(), l.data(), l.size());
+  tmp_cpu.set_order(ws.output_order());
+  tmp_gpu.set_order(ws.output_order());
+  tmp_gpu.Copy(tmp_cpu);
+
+  RepeatTensor(out, tmp_gpu, ws.GetInputBatchSize(0));
+}
+
+template <typename Backend>
+auto GetProperty<Backend>::GetPropertyReader(std::string_view key) -> PropertyReader {
+  if (key == "source_info") {
```

Comment:
This if-else is fine for a couple (2-3) of properties. If we expect to have more of these I would recommend using different approach - e.g. unordered_map.

Reply:
Realistically, I'd expect maybe 5 qualitatively different properties. If we have anything more, it'll likely be standardized somehow and likely not even hardcoded.
Having said that - this function is called only on operator's construction - so it's not on a critical path.